File: src/token_saver/filters/build_tools.py
"""Filters for build, package management, container, and infrastructure tool outputs."""

from __future__ import annotations

import re


def filter_npm_yarn(output: str) -> str:
    """Remove individual package download lines, keep summary and warnings/errors."""
    lines = output.split('\n')
    filtered = []
    for line in lines:
        if (
            re.match(r'^(npm |yarn |pnpm ).*(fetch|add|install)', line, re.IGNORECASE)
            or re.match(r'^\s*\[[0-9/]+\]\s+', line)
            or "GET" in line
            or "fetch" in line.lower()
            or "download" in line.lower()
        ):
            continue
        filtered.append(line)
    return '\n'.join(filtered)


def filter_cargo(output: str) -> str:
    """Remove successful compilation lines, keep warnings and errors."""
    lines = output.split('\n')
    filtered = []
    for line in lines:
        if re.match(r'^\s*Compiling\s+', line) or re.match(r'^\s*Downloaded\s+', line):
            continue
        filtered.append(line)
    return '\n'.join(filtered)


def filter_docker(output: str) -> str:
    """Filter verbose docker build/compose logs, keeping step headers, warnings, and errors."""
    lines = output.split('\n')
    filtered = []
    for line in lines:
        stripped = line.strip()
        # Skip intermediate hash lines and cache notes
        if re.match(r'^\s*--->\s+[a-f0-9]+', stripped) or stripped == "---> Using cache":
            continue
        if stripped.startswith("Removing intermediate container"):
            continue
        filtered.append(line)
    return '\n'.join(filtered)


def filter_maven_gradle(output: str) -> str:
    """Filter Maven and Gradle build output, removing repetitive download/task logs."""
    lines = output.split('\n')
    filtered = []
    for line in lines:
        stripped = line.strip()
        # Skip downloading / downloaded artifacts
        if stripped.startswith("Downloading from") or stripped.startswith("Downloaded from"):
            continue
        if stripped.startswith("Download http"):
            continue
        # Skip intermediate task execution progress
        if re.match(r'^>\s*Task\s+:[a-zA-Z0-9_:]+\s+UP-TO-DATE$', stripped):
            continue
        filtered.append(line)
    return '\n'.join(filtered)


def filter_generic_build(output: str) -> str:
    """Remove progress indicators, keep errors."""
    lines = output.split('\n')
    filtered = []
    for line in lines:
        if re.match(r'^[\[\d+%\]|\.]', line.strip()) and len(line) < 20:
            continue
        filtered.append(line)
    return '\n'.join(filtered)


def detect_and_filter_build(output: str) -> str | None:
    """Auto-detects the build tool and applies the right filter."""
    if "npm" in output or "yarn" in output or "pnpm" in output or "node_modules" in output:
        return filter_npm_yarn(output)
    elif ("Compiling" in output and "Cargo" in output) or "cargo" in output.lower():
        return filter_cargo(output)
    elif "Step " in output and ("--->" in output or "docker" in output.lower()):
        return filter_docker(output)
    elif "[INFO] Building" in output or "BUILD SUCCESS" in output or "BUILD FAILED" in output or "Gradle" in output:
        return filter_maven_gradle(output)
    return None

File: crates/token-saver-core/src/filters/build_tools.rs
//! Build, package manager, and container output filters in Rust.
//!
//! Prunes npm/yarn/pnpm download noise, docker layer hashes, and build logs.

use regex::Regex;

/// Removes individual package download lines, keeping summary and warnings/errors.
pub fn filter_npm_yarn(output: &str) -> String {
    let re_fetch = Regex::new(r"(?i)^(npm |yarn |pnpm ).*(fetch|add|install)").unwrap();
    let re_progress = Regex::new(r"^\s*\[[0-9/]+\]\s+").unwrap();

    let mut filtered = Vec::new();
    for line in output.lines() {
        let lower = line.to_lowercase();
        if re_fetch.is_match(line)
            || re_progress.is_match(line)
            || line.contains("GET ")
            || lower.contains("fetch")
            || lower.contains("download")
        {
            continue;
        }
        filtered.push(line);
    }
    filtered.join("\n")
}

/// Removes docker build intermediate container noise and cache hints.
pub fn filter_docker(output: &str) -> String {
    let re_layer = Regex::new(r"^\s*--->\s+[a-f0-9]+").unwrap();

    let mut filtered = Vec::new();
    for line in output.lines() {
        let stripped = line.trim();
        if re_layer.is_match(stripped)
            || stripped == "---> Using cache"
            || stripped.starts_with("Removing intermediate container")
        {
            continue;
        }
        filtered.push(line);
    }
    filtered.join("\n")
}

/// Auto-detects build/package manager type and prunes verbose lines.
pub fn detect_and_filter_build(output: &str) -> Option<String> {
    if output.contains("npm ")
        || output.contains("yarn ")
        || output.contains("pnpm ")
        || output.contains("node_modules")
    {
        Some(filter_npm_yarn(output))
    } else if output.contains("Step ")
        && (output.contains("--->") || output.to_lowercase().contains("docker"))
    {
        Some(filter_docker(output))
    } else {
        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_filter_npm_yarn() {
        let input = "npm WARN deprecated package@1.0.0\n\
npm fetch GET 200 https://registry.npmjs.org/package\n\
[1/4] Resolving packages...\n\
added 52 packages in 2.3s";

        let result = filter_npm_yarn(input);
        assert!(result.contains("npm WARN deprecated"));
        assert!(!result.contains("fetch GET 200"));
        assert!(!result.contains("[1/4] Resolving"));
        assert!(result.contains("added 52 packages in 2.3s"));
    }
}
